File: trunk/Consuming Services/Retrieve Metadata/MetadataHelper/MetadataHelper.fs
#light
open System
open System.ServiceModel
open System.ServiceModel.Channels
open System.ServiceModel.Description
open System.Text.RegularExpressions

module MetadataHelper =

    let internal messageMultiplier = 5
    
    let internal setMessageSize (el: TransportBindingElement) =
        el.MaxReceivedMessageSize <- el.MaxReceivedMessageSize * (int64 messageMultiplier)
        
    let internal (|NetTcp|NetPipe|Http|Https|) = function
        | "net.tcp"  -> NetTcp (new TcpTransportBindingElement())
        | "net.pipe" -> NetPipe (new NamedPipeTransportBindingElement())
        | "http"     -> Http (new HttpTransportBindingElement())
        | "https"    -> Https (new HttpsTransportBindingElement())
        | _          -> failwith "invalid scheme"

    let internal httpGetAddr (addr: string) =
        if Regex.IsMatch("?wsdl$", addr, RegexOptions.IgnoreCase)
            then addr
            else addr + "?wsdl"

    let internal serviceName<'C> =
        let t = typeof<'C>
        let attributes = t.GetCustomAttributes(typeof<ServiceContractAttribute>, false)
        
        if not t.IsInterface then None
        else if attributes.Length = 0 then None
        else
            let attr = attributes.[0] :?> ServiceContractAttribute
            match attr.Name,attr.Namespace with
            | null,null -> Some (t.ToString(), "http://tempuri.org")
            | null,ns   -> Some (t.ToString(), ns)
            | n,null    -> Some (n, "http://tempuri.org")
            | n,ns      -> Some (n, ns)
            
    let queryHttpEndpoint (mexAddress: string) (el: BindingElement) =
        let addr = httpGetAddr mexAddress
        let binding = new CustomBinding([| el |])
        let mexClient = new MetadataExchangeClient(binding)
        let metadata = mexClient.GetMetadata(new Uri(addr), MetadataExchangeClientMode.HttpGet)
        let importer = new WsdlImporter(metadata)
        
        importer.ImportAllEndpoints()

    let tryQueryHttpEndpoint (mexAddress: string) (el: BindingElement) =
        try
            Some (queryHttpEndpoint mexAddress el)
        with _ -> None
    
    let queryMexEndpoint (mexAddress: string) (bindingElement: BindingElement) =
        let binding = new CustomBinding([| bindingElement |])
        let mexClient = new MetadataExchangeClient(binding)
        let metadata = mexClient.GetMetadata(new EndpointAddress(mexAddress))
        let importer = new WsdlImporter(metadata)
        
        importer.ImportAllEndpoints()
    
    let tryQueryMexEndpoint (mexAddress: string) (bindingElement: BindingElement) =
        try
            Some (queryMexEndpoint mexAddress bindingElement)
        with _ -> None

    let getEndpoints (mexAddress: string) =
        let address = new Uri(mexAddress)
        
        match address.Scheme with
        | NetTcp el  ->
            setMessageSize el
            tryQueryMexEndpoint mexAddress el
        | NetPipe el ->
            setMessageSize el
            tryQueryMexEndpoint mexAddress el
        | Http el    ->
            setMessageSize el
            match tryQueryMexEndpoint mexAddress el with
            | None -> tryQueryHttpEndpoint mexAddress el
            | es   -> es
        | Https el   ->
            setMessageSize el
            match tryQueryMexEndpoint mexAddress el with
            | None -> tryQueryHttpEndpoint mexAddress el
            | es   -> es
    
    let queryContract (mexAddress: string) (contractNamespace: string) (contractName: string) =
       
        let matches (endpoint: ServiceEndpoint) =
            endpoint.Contract.Namespace = contractNamespace &&
            endpoint.Contract.Name = contractName
    
        if String.IsNullOrEmpty(contractNamespace) then false
        else if String.IsNullOrEmpty(contractName) then false
        else
            match getEndpoints mexAddress with
            | None    -> false
            | Some es ->
                es
                |> Seq.exists matches

    let queryContractType<'C> (mexAddress: string) =
        match serviceName<'C> with
        | None        -> false
        | Some (n,ns) -> queryContract mexAddress n ns

File: trunk/CodeRunner/ServiceModel.Extensions/MetadataHelper.cs
using System.ServiceModel.Channels;
using System.ServiceModel.Description;

namespace System.ServiceModel.Extensions
{
    public static class MetadataHelper
    {
        const int MessageSizeMultiplier = 5;

        static ServiceEndpointCollection GetEndpoints(
            Uri mexUri,
            BindingElement bindingElement)
        {
            CustomBinding binding = new CustomBinding(bindingElement);

            MetadataExchangeClient MEXClient = new MetadataExchangeClient(binding);
            MetadataSet metadata = MEXClient.GetMetadata(new EndpointAddress(mexUri));

            MetadataImporter importer = new WsdlImporter(metadata);
            return importer.ImportAllEndpoints();
        }

        public static ServiceEndpointCollection GetEndpoints(string mexAddress)
        {
            Uri mexUri = new Uri(mexAddress);
            TransportBindingElement transportElement;
            switch (mexUri.Scheme)
            {
                case "net.tcp":
                    transportElement = new TcpTransportBindingElement();
                    break;
                case "net.pipe":
                    transportElement = new NamedPipeTransportBindingElement();
                    break;
                case "http":
                    transportElement = new HttpTransportBindingElement();
                    break;
                case "https":
                    transportElement = new HttpsTransportBindingElement();
                    break;
                default:
                    throw new NotSupportedException("Scheme not supported.");
            }
            transportElement.MaxReceivedMessageSize *= MessageSizeMultiplier;
            return GetEndpoints(mexUri, transportElement); ;
        }

        public static bool SupportsContract(
            string mexAddress,
            Type contractType)
        {
            if (contractType.IsInterface == false)
            {
                throw new ArgumentException(contractType.Name + " is not an interface", "contractType");
            }
            object[] attributes = contractType.GetCustomAttributes(typeof(ServiceContractAttribute), false);
            if (attributes.Length == 0)
            {
                throw new ArgumentException("Interface does not have the ServiceContractAttribute", "contractType");
            }
            ServiceContractAttribute attribute = attributes[0] as ServiceContractAttribute;
            if (attribute.Name == null)
            {
                attribute.Name = contractType.Name;
            }
            if (attribute.Namespace == null)
            {
                attribute.Namespace = "http://tempuri.org/";
            }
            return SupportsContract(mexAddress, attribute.Namespace, attribute.Name);
        }

        public static bool SupportsContract(
            string mexAddress,
            string contractNamespace,
            string contractName)
        {
            if (string.IsNullOrEmpty(contractNamespace))
            {
                throw new ArgumentException("Empty namespace", "contractNamespace");
            }
            if (string.IsNullOrEmpty(contractName))
            {
                throw new ArgumentException("Empty name", "contractName");
            }
            ServiceEndpointCollection endpoints = GetEndpoints(mexAddress);
            foreach (ServiceEndpoint endpoint in endpoints)
            {
                if (endpoint.Contract.Namespace == contractNamespace &&
                    endpoint.Contract.Name == contractName)
                {
                    return true;
                }
            }
            return false;
        }
    }
}
